pipeline-model-declarative-agent/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgentDescriptor.java

```diff
@@ -0,0 +1,190 @@
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Descriptor for {@link DeclarativeAgent}.
+ *
+ * @author Andrew Bayer
+ */
+public abstract class DeclarativeAgentDescriptor extends Descriptor<DeclarativeAgent> {
+
+    /**
+     * The name for this agent type. Should match the {@code Symbol} on the class.
+     *
+     * @return The name.
+     */
+    public abstract @Nonnull String getName();
+
+    /**
+     * The full package and class name for the {@link DeclarativeAgentScript} class corresponding to this.
+     *
+     * @return The class name.
+     */
+    public abstract @Nonnull String getDeclarativeAgentScriptClass();
+
+    /**
+     * Creates an instance of the corresponding {@link DeclarativeAgent} from the given arguments.
+     *
+     * @param arguments A map of strings/objects to be passed to the constructor.
+     * @return An instantiated {@link DeclarativeAgent}
+     * @throws Exception
+     */
+    public DeclarativeAgent newInstance(Map<String,Object> arguments) throws Exception {
+        return new DescribableModel<>(clazz).instantiate(arguments);
+    }
+
+    public static ExtensionList<DeclarativeAgentDescriptor> all() {
+        return ExtensionList.lookup(DeclarativeAgentDescriptor.class);
+    }
+
+    /**
+     * Get a map of name-to-{@link DescribableModel} of all known/registered descriptors.
+     *
+     * @return A map of name-to-{@link DescribableModel}s
+     */
+    public static Map<String,DescribableModel> getDescribableModels() {
+        Map<String,DescribableModel> models = new HashMap<>();
+
+        for (DeclarativeAgentDescriptor d : getOrderedDescriptors()) {
+            Set<String> symbolValues = SymbolLookup.getSymbolValue(d);
+
+            if (!symbolValues.isEmpty()) {
+                models.put(symbolValues.iterator().next(), new DescribableModel<>(d.clazz));
+            }
+
+        }
+
+        return models;
+    }
+
+    /**
+     * Get the map of the subset of descriptors with single-arguments - i.e., "none" and "any".
+     * @return A map of descriptors with just one argument.
+     */
+    public static Map<String,DescribableModel> singleArgModels() {
+        Map<String,DescribableModel> models = new HashMap<>();
+
+        for (Map.Entry<String,DescribableModel> entry : getDescribableModels().entrySet()) {
+            if (entry.getValue().getParameters().isEmpty()) {
+                models.put(entry.getKey(), entry.getValue());
+            }
+        }
+
+        return models;
+    }
+
+    /**
+     * An ordered (by Extension ordinal, in descending order) list of all descriptors.
+     * @return A list of descriptors.
+     */
+    public static List<DeclarativeAgentDescriptor> getOrderedDescriptors() {
+        List<DeclarativeAgentDescriptor> orderedDescriptors = new ArrayList<>();
+
+        List<ExtensionComponent<DeclarativeAgentDescriptor>> extensionComponents = new ArrayList<>(all().getComponents());
+        Collections.sort(extensionComponents);
```

Comment:
Delete this method. `ExtensionList` does this automatically, so `all()` should suffice.